include/SampleBuffer.h

```diff
@@ -26,385 +26,396 @@
-	{
-		return "samplebuffer";
-	}
-	virtual void saveSettings(QDomDocument& doc, QDomElement& _this ) override;
-	virtual void loadSettings(const QDomElement& _this ) override;
-
-	bool play( sampleFrame * _ab, handleState * _state,
-				const fpp_t _frames,
-				const float _freq,
-				const LoopMode _loopmode = LoopOff );
+	/**
+	 * Load this sample buffer from @a _audio_file,
+	 * @param _audio_file	path to an audio file.
+	 */
+	explicit SampleBuffer(const QString &_audio_file, bool ignoreError=false);
 
-	void visualize( QPainter & _p, const QRect & _dr, const QRect & _clip, f_cnt_t _from_frame = 0, f_cnt_t _to_frame = 0 );
-	inline void visualize( QPainter & _p, const QRect & _dr, f_cnt_t _from_frame = 0, f_cnt_t _to_frame = 0 )
-	{
-		visualize( _p, _dr, _dr, _from_frame, _to_frame );
-	}
+	/**
+	 * Load the sample buffer from a base64 encoded string.
+	 * @param base64Data	The data.
+	 * @param sample_rate	The data's sample rate.
+	 */
+	SampleBuffer(const QString &base64Data, sample_rate_t sample_rate);
 
-	QPair<QPolygonF, QPolygonF>  visualizeToPoly( const QRect & _dr, const QRect & _clip, f_cnt_t _from_frame = 0, f_cnt_t _to_frame = 0) const;
+	SampleBuffer(DataVector &&movedData, sample_rate_t sampleRate);
 
-	inline void resetAudioFile()
-	{
-		m_audioFile = "";
+	inline virtual QString nodeName() const override {
+		return "samplebuffer";
 	}
 
-	inline const QString & audioFile() const
+	SampleBuffer(SampleBuffer &&sampleBuffer) noexcept
+			:
+			m_data{std::move(sampleBuffer.m_data)},
+			m_playInfo(std::move(sampleBuffer.m_playInfo))
 	{
-		return m_audioFile;
+		emit sampleUpdated(UpdateType::Clear);
 	}
 
-	inline f_cnt_t startFrame() const
-	{
-		return m_startFrame;
-	}
+	SampleBuffer &operator=(SampleBuffer &&other) noexcept {
+		DataChangeHelper helper(this, UpdateType::Clear);
 
-	inline f_cnt_t endFrame() const
-	{
-		return m_endFrame;
-	}
+		m_data = std::move(other.m_data);
+		m_playInfo = std::move(other.m_playInfo);
 
-	inline f_cnt_t loopStartFrame() const
-	{
-		return m_loopStartFrame;
+		return *this;
 	}
 
-	inline f_cnt_t loopEndFrame() const
-	{
-		return m_loopEndFrame;
-	}
+	virtual void saveSettings(QDomDocument &doc, QDomElement &_this) override;
 
-	void setLoopStartFrame( f_cnt_t _start )
-	{
-		m_loopStartFrame = _start;
-	}
+	virtual void loadSettings(const QDomElement &_this) override;
 
-	void setLoopEndFrame( f_cnt_t _end )
-	{
-		m_loopEndFrame = _end;
-	}
+	bool play(sampleFrame *_ab, handleState *_state,
+			  const fpp_t _frames,
+			  const float _freq,
+			  const LoopMode _loopmode = LoopMode::LoopOff);
 
-	void setAllPointFrames( f_cnt_t _start, f_cnt_t _end, f_cnt_t _loopstart, f_cnt_t _loopend )
-	{
-		m_startFrame = _start;
-		m_endFrame = _end;
-		m_loopStartFrame = _loopstart;
-		m_loopEndFrame = _loopend;
+	template<class LaunchType=LaunchType::Async>
+	void setLoopStartFrame(f_cnt_t _start) {
+		runToSetPlayInfo<LaunchType>([=](GuardedPlayInfo &playInfo) {
```

Comment:
To fix build errors for Windows, you should either move bodies of this kind of functions to the `.cpp` file or add `LMMS_EXPORT` to `internal::SampleBufferPlayInfo` and `internal::SampleBufferData`.